# Lecture 8.1 - Error Handling / Exception
So far, we have found on multiple occasions that our program crashes. Common issues that causes crashes are syntax errors, type mismatch or visibility errors. Crashes are something we definitely don't want happening in our finished software. Take for example an input function that takes input from the user via a Scanner. The input needs to be in `int` form, but the user supplies a `double`. In this case, the program crashes, as there's a type mismatch between the provided input (`double`) and the input expected by the Scanner object (`int`). 

Errors can be found both at compilation (if the program does not compile properly) or at runtime (e.g. if the user provides faulty input). A compile-time error will be detected when the java-file is compiled, and a runtime error will be deteted by the JVM once the program is running. In both of these cases, an `Exception` object is created, containg information on the type of error, the java-file that produced it, and the state of the program when it occurred. Returning an `Exception` object in this manner is called <i>throwing</i> an exception.

`Exception` is a class (just like any other Java class), and consequently, an `Exception`-object is an object (just like any other object in Java) which is created when an exception is thrown. The `Exception` class itself is usually not the one used in error handling, but rather, it is extended (via the `extends` keyword) into a more specific error class, e.g. `InputMismatchException` or `ArrayIndexOutOfBoundsException`. 

### The Call Stack and Catching exceptions
In order to understand how an exception is thrown and handled, we need to introduce the <i>call stack</i>. The call stack is the list of consequitive methods called when the error occured. If our error is produced in a function called `multiplication()`, then the call stack will consist of [`multiplication()`] - [`main()`]. Generally, one can describe the call stack with the following schedule: [`method a (where the exception was thrown)`] - [`method b`] - [`method c`] - [`main()`]. If an exception is thrown, the exception handler will search through the call stack until it finds a code block that can handle the exception. Handling an exception is called to <i>catch</i> the exception.

![callstack](/assets/lecture_8/callstack.PNG)

### Three types of exceptions
Exceptions can be categorized into three different categories:
<ul>
<li>Checked exceptions</li>
<li>Errors</li>
<li>Unchecked (Runtime) exceptions</li>
</ul><br>

<i>Checked exceptions</i> are the most commonly occuring, and relate to types of errors that a software should be able to anticipiate, and therefore provide error handling for. Examples include e.g. if the user provides a filename that doesn't exist (`FileNotFoundException`). A checked exception must be handled by an error handling codeblock, else the code will not compile.

<i>Errors</i> are external to the software, and are caused by e.g. malfunctioning hardware, network issues, or similar. These can also be handled, but conversely to checked exceptions, they do not have to be in order for the program to compile properly.

Finally, <i>Unchecked Exception</i> (also called <i>Runtime Exception</i>) is an internal error that the program will not be able to recover from, even with proper error handling. These usually indicate programming bugs, such as logic errors or improper use of an API. Examples include `NullPointerException`, `ArrayIndexOutOfBoundsException` or `InputMismatchException`. An unchecked exception does not need to be handled, but instead usually indicate some kind of logical problem with the code which probably warrants being taken care of.

